modin/pandas/groupby.py

```diff
@@ -128,7 +129,46 @@ def sem(self, ddof=1):
         return self._default_to_pandas(lambda df: df.sem(ddof=ddof))
 
     def mean(self, *args, **kwargs):
-        return self._apply_agg_function(lambda df: df.mean(*args, **kwargs))
+        fallback = False
+        converted_columns = {}
+        numeric_only = kwargs.get("numeric_only", False)
+        if not numeric_only:
+            for col, dt in self._query_compiler.dtypes.items():
+                if is_datetime64_any_dtype(dt):
+                    if self._df[col].hasnans:
+                        fallback = True
+                        break
+                    else:
+                        converted_columns[col] = dt
```

Comment:
in order to minimize backend-API calls I would replace it with something like this:
```python
datetime_cols = self._df[[col for col, dt in self._df.dtypes.items() if is_datetime64_any_dtype(dt)]]
has_nans_mask = datetime_cols.isna().any()
```
I've compared these two approaches and it appeared that the proposed one almost always faster:

<details><summary>Micro-benchmark and its results</summary>

```python
import modin.pandas as pd
import numpy as np
import timeit


def is_dtype_to_pick(dtype):
    return dtype == "float"


def fn1(df):
    to_convert = {}
    fallback = False
    for col, dt in df.dtypes.items():
        if is_dtype_to_pick(dt):
            if df[col].hasnans:
                fallback = True
                break
            else:
                to_convert[col] = dt


def fn2(df):
    cols_to_select = [col for col, dt in df.dtypes.items() if is_dtype_to_pick(dt)]
    if len(cols_to_select) == 0:
        return
    selected_df = df[cols_to_select]
    fallback = selected_df.isna().any().any()
    if not fallback:
        to_convert = selected_df.dtypes.items()


NROWS = 1_000_000
NCOLS = 10

df_no_floats = pd.DataFrame({f"col{i}": np.arange(NROWS) for i in range(NCOLS)})
df_has_half_float_no_nans = df_no_floats.astype({col: "float" for col in df_no_floats.columns[:len(df_no_floats.columns) // 2]})

df_has_half_float_nans_first = df_has_half_float_no_nans.copy()
df_has_half_float_nans_first.loc[0, df_has_half_float_nans_first.select_dtypes("float").columns[0]] = np.nan

df_has_half_float_nans_last = df_has_half_float_no_nans.copy()
df_has_half_float_nans_first.loc[0, df_has_half_float_nans_first.select_dtypes("float").columns[-1]] = np.nan

NRUNS = 5

print("df_no_float:")
print("for-loop:", timeit.timeit(lambda: fn1(df_no_floats), number=NRUNS))
print("non-for-loop:", timeit.timeit(lambda: fn2(df_no_floats), number=NRUNS))

print("\ndf_has_half_float_no_nans:")
print("for-loop:", timeit.timeit(lambda: fn1(df_has_half_float_no_nans), number=NRUNS))
print("non-for-loop:", timeit.timeit(lambda: fn2(df_has_half_float_no_nans), number=NRUNS))

print("\ndf_has_half_float_nans_first:")
print("for-loop:", timeit.timeit(lambda: fn1(df_has_half_float_nans_first), number=NRUNS))
print("non-for-loop:", timeit.timeit(lambda: fn2(df_has_half_float_nans_first), number=NRUNS))

print("\ndf_has_half_float_nans_last:")
print("for-loop:", timeit.timeit(lambda: fn1(df_has_half_float_nans_last), number=NRUNS))
print("non-for-loop:", timeit.timeit(lambda: fn2(df_has_half_float_nans_last), number=NRUNS))

```

Results:
```
df_no_float:
for-loop: 8.46749753691256e-05
non-for-loop: 6.431201472878456e-05

df_has_half_float_no_nans:
for-loop: 5.488216175988782
non-for-loop: 1.856114284018986

df_has_half_float_nans_first:
for-loop: 1.3251252140034921
non-for-loop: 1.966320100997109

df_has_half_float_nans_last:
for-loop: 5.407154903979972
non-for-loop: 1.9547889120294712

```

</details>

Reply:
the `datetime_cols = ...` here would make a copy in pandas (not sure about modin).  I suggest `select_dtypes`